### PR TITLE
Backend plugins: Support Forward OAuth Identity for backend data source plugins

### LINKED ISSUE #26023: Backend plugins that authenticate users to their data store need OAuth2 token propagated
**What would you like to be added:**

Request: If configured to do so, pass the oauth token to the backend datasource code.

**Why is this needed:**

Datasources that authenticate users to their data store need access to the identity of the Grafana user.

(Timely = https://github.com/NationalSecurityAgency/timely)

Currently, we are using a front-end app that authenticates the user and authenticates them to Grafana using AuthProxy.  After that, the datasource requests come from the browser and use PKI client certificates to authenticate to the backend data store.   

Now that we are implementing a backend datasource (to use the alerting feature), we are looking into using OAuth2 with a custom OAuth2 provider (i.e. generic for Grafana config purposes) for authenticating to Grafana and to Timely (the data store).

While backend-proxied datasources have the capability of forwarding the oauth token (oauthPassThru), this capability does not exist for backend datasource plugins.  Currently only the context and request are passed across the RPC wire.

### PR DESCRIPTION
**Pull request by billoley**
fixes #26023 

If configured to do so, pass the oauth token to the backend datasource code.
This is already being done in ds_proxy.go for backend proxied datasources.
This change moves the code from the addOAuthPassThruAuth method to a new package oauthtoken with slight refactoring.
This allows both datasource_plugin_wrapper_v2.go and ds_proxy.go to use the same code to get an up-to-date token.

### CODE REVIEW

**Comment by billoley** [comment]
Code flow through social.GetCurrentOAuthToken is already tested by ds_proxy_test.go when getDirector() is called with oauth configured

**Comment by billoley** [comment]
Should this PR be rebased on branch v7.1.x instead of master.... which will be 7.2.0?

File: pkg/login/social/social.go
Comment (kminehart) [inline pkg/login/social/social.go:237]:
```suggestion
func GetOAuthHttpClient(name string) (*http.Client, error) {
```

File: pkg/login/social/social.go
Comment (kminehart) [inline pkg/login/social/social.go:246]:
```suggestion
	info, ok := setting.OAuthService.OAuthInfos[name]
	if !ok {
		return nil, fmt.Errorf("Could not find %s in OAuth Settings", name)
	}
	
	// handle call back
	tr := &http.Transport{
		Proxy: http.ProxyFromEnvironment,
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: info.TlsSkipVerify,
		},
	}
```

If `name` is not in the `OAuthInfos` map, trying to access `TlsSkipVerify` will panic.

File: pkg/login/social/social.go
Comment (kminehart) [inline pkg/login/social/social.go:252]:
```suggestion
	if info.TlsClientCert != "" || info.TlsClientKey != "" {
		cert, err := tls.LoadX509KeyPair(info.TlsClientCert, info.TlsClientKey)
```

File: pkg/login/social/social.go
Comment (kminehart) [inline pkg/login/social/social.go:261]:
```suggestion
	if info.TlsClientCa != "" {
		caCert, err := ioutil.ReadFile(info.TlsClientCa)
```

File: pkg/login/social/social.go
Comment (kminehart) [inline pkg/login/social/social.go:272]:
```suggestion
func GetCurrentOAuthToken(ctx context.Context, user models.SignedInUser) (*oauth2.Token, error) {
```

File: pkg/login/social/social.go
Comment (kminehart) [inline pkg/login/social/social.go:273]: This tells me that you did not run `go fmt` :P

File: pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go
Comment (kminehart) [inline pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:69]:
```suggestion
			}
			delete(query.Headers, "Authorization")
			query.Headers["Authorization"] = fmt.Sprintf("%s %s", token.Type(), token.AccessToken)
```

I think in most cases, and `social.GetCurrentOauthToken` included, it's fair to assume that if `err` is nil, then the other value in the response is not.

File: pkg/api/pluginproxy/ds_proxy.go
Comment (kminehart) [inline pkg/api/pluginproxy/ds_proxy.go:219]: If `err` is nil then it's very fair to assume that `token` is not.

```suggestion
			req.Header.Del("Authorization")
			req.Header.Add("Authorization", fmt.Sprintf("%s %s", token.Type(), token.AccessToken))
```

**Review by kminehart — DISMISSED** [review]
I requested some minor style changes. I am much less familiar with Grafana than the listed reviewers so I would still like them to review this.

This would be super useful for users of the github plugin (https://github.com/grafana/grafana-github-datasource)

**Comment by kminehart** [comment]
It would be nice if, instead of configuring this per OAuth provider, we could configure this per datasource.

**Comment by billoley** [comment]
> It would be nice if, instead of configuring this per OAuth provider, we could configure this per datasource.

Passing the oauth token through _is_ configured on a per datasource basis for both backand plugins and for the proxied plugins

Backend
https://github.com/grafana/grafana/pull/27055/files#diff-71d6e3ee71ac498cfe8fcd0e3ee9a965R60

Proxied
https://github.com/grafana/grafana/pull/27055/files#diff-cd5947c41782508115b4ac9e25218aa1R210

File: pkg/login/social/social.go
Comment (billoley) [reply pkg/login/social/social.go:278]: I thought I ran the required linting and formatting.  I will take a look.

**Comment by kminehart** [comment]
Oh! For some reason I assumed that it was per datasource. Sorry about that!

**Comment by billoley** [comment]
@kminehart - I'm pretty sure that I made all requested changes and then rebased on master to get the ci build to complete.  One of your requested change (that I think I made) is blocking the PR for some reason.

Also - Any idea if someone can review this PR to get it into 7.2?

**Review by marefr — CHANGES_REQUESTED** [review]
Thanks. In general looks good. There's a few minor things I would like to see changed, see comments.

The biggest problem right now is that the social package have to much responsibility since it now besides handling OAuth providers and provide OAuth mechanism it also makes database calls. 

To me it seems like we're lacking a package that could coordinate receiving of stored OAuth token, refresh it if needed and store the updated token in the database. This package could then be used by ds_proxy.go, datasource_plugin_wrapper_v2.go  and potentially other code since backend plugins also has support for resources and potentially other capabilities in the future.

Question is just where to put this package and how to name it. My suggestions that comes to mind:
- pkg/services/oauth
- pkg/services/oauthtoken

Thoughts/questions?

**Comment by billoley** [comment]
> Thanks. In general looks good. There's a few minor things I would like to see changed, see comments.
> 
> The biggest problem right now is that the social package have to much responsibility since it now besides handling OAuth providers and provide OAuth mechanism it also makes database calls.
> 
> To me it seems like we're lacking a package that could coordinate receiving of stored OAuth token, refresh it if needed and store the updated token in the database. This package could then be used by ds_proxy.go, datasource_plugin_wrapper_v2.go and potentially other code since backend plugins also has support for resources and potentially other capabilities in the future.
> 
> Question is just where to put this package and how to name it. My suggestions that comes to mind:
> 
> * pkg/services/oauth
> * pkg/services/oauthtoken
> 
> Thoughts/questions?

I chose the social package as it seemed most the most appropriate from existing packages, but figured that the code owners might have other ideas.  I think either suggestion would be fine, but probably oauthtoken to keep it separate from the Grafana code that implements an oauth provider.

File: pkg/login/social/social.go
Comment (billoley) [reply pkg/login/social/social.go:293]: I saw an http ssl error on getting a new token (refresh) and then found that login_oauth.go was setting the http client into the context.  It made sense that the same thing should be done for refreshing the token.

File: pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go
Comment (billoley) [reply pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:60]: I think this helper method would fit well in pkg/services/oauthtoken since both datasource_plugin_wrapper_v2.go and ds_proxy.go will be accessing the same package to get the current OAuthToken.

File: pkg/login/social/social.go
Comment (billoley) [reply pkg/login/social/social.go:279]: Would the same code be fine if it is in the new oauthtoken package as you suggested?

If GetCurrentOAuthToken (in oauthtoken, not social) has to take this as an argument, then we will have the database code twice -- once in datasource_plugin_wrapper_v2.go and once in ds_proxy.go.

File: pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go
Comment (marefr) [reply pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:60]: Yeah makes sense. I wrote that comment before realizing that things probably should be moved around.

File: pkg/login/social/social.go
Comment (marefr) [reply pkg/login/social/social.go:279]: Yes, parts of it. Would make sense to have database interactions in oauthtoken package. oauthtoken package should use the social package for retrieving the token (and refresh if needed) since the social package has all the internal knowledge of providers, http clients etc. Basically the oauthtoken package will be the coordinator for getting a token, but reach out to other packages for getting the user token from database, refresh it, if needed, and store it in database if refreshed. 

Does this make sense?

File: pkg/login/social/social.go
Comment (billoley) [reply pkg/login/social/social.go:293]: Moved the OAuth HTTP Client changes to a new PR -> https://github.com/grafana/grafana/pull/27523
It will be a lot easier to include that new call on this PR once that PR is approved and merged.
I will back that code out of this PR (except for calling it once it's merged in)

Comment (marefr) [reply pkg/login/social/social.go:293]: Great. I have reviewed #27523

**Comment by billoley** [comment]
Sorry for the force push.  Between splitting the PR and rebasing a bunch of prior commits, this seemed cleaner (and easier)

**Comment by marefr** [comment]
Thanks. Planning to review this again soon, hopefully today or else beginning next week. Have had a lot of other things going on.

File: pkg/login/social/social.go
Comment (billoley) [inline pkg/login/social/social.go:260]: Moved this fix to another PR https://github.com/grafana/grafana/pull/27804 since it is a bug instead of an improvement related to this task.

**Comment by billoley** [comment]
> Thanks. Planning to review this again soon, hopefully today or else beginning next week. Have had a lot of other things going on.

Understand.  Thanks for your time.

**Comment by marefr** [comment]
Sorry for keeping you waiting. Will make sure to review it beginning next week.

**Comment by billoley** [comment]
Any update on review?

**Comment by aknuds1** [comment]
We're supposed to review it tomorrow.

**Comment by marefr** [comment]
Reviewing this now. Really sorry for keeping you waiting @billoley

**Review by marefr — CHANGES_REQUESTED** [review]
Great work. Some minor things left related to logging/error handling.

File: pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go
Comment (billoley) [reply pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:62]: I agree that moving the user nil check into oauthtoken.GetCurrentOAuthToken makes sense.  If there is no user, then there is no token and we can return nil, nil.

I disagree with moving the IsOAuthPassThruEnabled(ds) check into GetCurrentOAuthToken.  As written, GetCurrentOAuthToken returns the token (possibly refreshed) if there is one.  The check on whether IsOAuthPassThruEnabled(ds) is true or false does not fit into the gereralized purpose of GetCurrentOAuthToken.  It seems wrong to me unless there is something I am missing.

However.... your code.   Let me know.

File: pkg/services/oauthtoken/oauth_token_util.go
Comment (aknuds1) [inline pkg/services/oauthtoken/oauth_token_util.go:30]:
```suggestion
		return nil, fmt.Errorf("failed to get OAuth connector error=%s", err)
```

File: pkg/services/oauthtoken/oauth_token_util.go
Comment (aknuds1) [inline pkg/services/oauthtoken/oauth_token_util.go:35]:
```suggestion
		return nil, fmt.Errorf("failed to create OAuth http client error=%s", err)
```

File: pkg/services/oauthtoken/oauth_token_util.go
Comment (aknuds1) [inline pkg/services/oauthtoken/oauth_token_util.go:60]:
```suggestion
			return nil, fmt.Errorf("failed to update auth info during token refresh, userId=%d username=%s error=%s", user.UserId, user.Login, err)
```

**Review by aknuds1 — COMMENTED** [review]
Generally LGTM, but spotted some nits.

File: pkg/services/oauthtoken/oauth_token_util.go
Comment (aknuds1) [inline pkg/services/oauthtoken/oauth_token_util.go:48]:
```suggestion
		return nil, fmt.Errorf("failed to retrieve access token from OAuth, provider=%s userid=%d username=%s error=%s", authInfoQuery.Result.AuthModule, user.UserId, user.Login, err)
```
```suggestion
		return nil, fmt.Errorf("Failed to retrieve access token from OAuth provider=%s userid=%d username=%s error=%s", authInfoQuery.Result.AuthModule, user.UserId, user.Login, err)
```
```suggestion
		return nil, fmt.Errorf("Failed to retrieve access token from OAuth provider=%s userid=%d username=%s error=%s", authInfoQuery.Result.AuthModule, user.UserId, user.Login, err)
```

**Review by aknuds1 — COMMENTED** [review]
Spotted a remaining nit.

File: pkg/services/oauthtoken/oauth_token_util.go
Comment (billoley) [reply pkg/services/oauthtoken/oauth_token_util.go:24]: At first, I was returning an error here, but found that when I was logged in as admin with username/password, the Dispatch call returned an error.  Should that be true or am I mistaken?  

If Dispatch returning an error in the non-oauth case is correct, then I think this should remain as-is returning nil, nil instead of propagating the error.

File: pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go
Comment (marefr) [reply pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go]: I was only referring to the nil check and not moving `IsOAuthPassThruEnabled`

File: pkg/services/oauthtoken/oauth_token_util.go
Comment (marefr) [reply pkg/services/oauthtoken/oauth_token_util.go:29]: Returning `nil, nil` I don't like. As a consumer and Go developer I'm used to checking the err for non-nil and if not expect other returned value is valid. With this you have to check both return values for nil. But I understand why that may be needed here. Returning a special error if user not found and letting the consumer check this is not good either.

Looking at the code handling the GetAuthInfoQuery seems like `models.ErrUserNotFound` is returned as error if no user is found. So I would suggest checking if `err == models.ErrUserNotFound` and if so debug log, otherwise error log. And keeping the `return nil, nil`.

Comment (billoley) [reply pkg/services/oauthtoken/oauth_token_util.go:29]: I understand checking models.ErrUserNotFound, logging at debug level, and still returning nil, nil in that case.
If there is an error that is not models.ErrUserNotFound, should I log that at error level and return the error (return nil, err)?

Comment (marefr) [reply pkg/services/oauthtoken/oauth_token_util.go:29]: > If there is an error that is not models.ErrUserNotFound, should I log that at error level and return the error (return nil, err)?

I was thinking log that at error level and return nil, nil to sort of mimic the old behavior. WDYT?

Comment (billoley) [reply pkg/services/oauthtoken/oauth_token_util.go:29]: The only place where the oauth token was being added in this way previously was in ds_proxy

if proxy.ds.JsonData != nil && proxy.ds.JsonData.Get("oauthPassThru").MustBool() {
	addOAuthPassThruAuth(proxy.ctx, req)
}

with the beginning of the addOAuthPassThruAuth function being:

func addOAuthPassThruAuth(c *models.ReqContext, req *http.Request) {
	authInfoQuery := &models.GetAuthInfoQuery{UserId: c.UserId}
	if err := bus.Dispatch(authInfoQuery); err != nil {
		logger.Error("Error fetching oauth information for user", "userid", c.UserId, "username", c.Login, "error", err)
		return
	}

To maintain the previous status quo, we have to *at some level* ignore *any* error that comes back from the GetAuthInfoQuery call to bus.Dispatch and carry on with sending the request without oauth. An error in that particular location would not stop the request.  This at least applies to the ds_proxy case since it existed previously.  To maintain backward compatibility and consistency, what you propose sounds fine to me.

Comment (billoley) [reply pkg/services/oauthtoken/oauth_token_util.go:33]: Committed that change to differentiate the log level

Comment (billoley) [reply pkg/services/oauthtoken/oauth_token_util.go:33]: Would it be possible to get this into the next 7.3.0 beta?

Comment (billoley) [reply pkg/services/oauthtoken/oauth_token_util.go:33]: Any idea why continuous-integration/drone/pr started failing on this PR?

**Comment by aknuds1** [comment]
I would try merging with latest master to fix CI.

File: pkg/services/oauthtoken/oauth_token_util.go
Comment (marefr) [reply pkg/services/oauthtoken/oauth_token_util.go:33]: Sorry for slow response. Great. Will have a look. Unfortunately it may be too late to get this into v7.3 stable, but will consider/discuss it.

**Comment by marefr** [comment]
Decided to change the `GetCurrentOAuthToken` to only return token. Makes a cleaner API to not returning an error and keep the logs of errors in the oauthtoken package for now. One error created multiple log messages earlier. Decided to continue the request in both data source proxy and backend plugin wrapper in case of an error and give the data source the responsibility of handling the case of a missing token/Authorization header. WDYT?

**Comment by billoley** [comment]
> Decided to change the `GetCurrentOAuthToken` to only return token. Makes a cleaner API to not returning an error and keep the logs of errors in the oauthtoken package for now. One error created multiple log messages earlier. Decided to continue the request in both data source proxy and backend plugin wrapper in case of an error and give the data source the responsibility of handling the case of a missing token/Authorization header.
> 
> @billoley WDYT?

I agree that it's cleaner returning just the token (or nil) and handling all errors with logging inside the GetCurrentOAuthToken method.  I like it.

**Review by marefr — APPROVED** [review]
LGTM

**Comment by marefr** [comment]
@billoley thank you for contributing to Grafana! Sorry it took some time to get this merged, but I really appreciate your patience and  attitude working together on this one.

**Pull request merged**
